### PR TITLE
opcodes for known sections

### PR DESCRIPTION
**Pull request by Krovatkin**
Per https://github.com/WebAssembly/design/pull/740
this adds a few bits to support opcodes for known sections 
fyi @commonlisp  @arunetm 

@cellulle @mikeholman homeless PR looks for a nice review and branch to be adopted into

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/1460)

### CODE REVIEW

File: test/wasm/f32_known_sections.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft Corporation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+const blob = WScript.LoadBinaryFile('f32_known_sections.wasm');
```
Comment (commonlisp) [inline test/wasm/f32_known_sections.js]: Can we use a simpler test case for this?

**Comment by commonlisp** [comment]
Looks like we have to rebase this one.

**Comment by commonlisp** [comment]
@Cellule Just want to put this on your radar since other changes (globals and table/elements) will depend on it.

**Comment by Cellule** [comment]
Reviewed 1 of 4 files at r1, 9 of 11 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 177 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1460#-KSEDZ_nswcEKaMz2rzd:-KSEDZ_nswcEKaMz2rze:bz0b7ir) ([raw file](https://github.com/microsoft/chakracore/blob/6d9b004f8e728e70a1432dce15e06a59dba65a79/lib/WasmReader/WasmBinaryReader.cpp#L177)):*
> ```cpp
>     else
>     {
>         sectionName = (char*)(m_pc); //including "name" section
> ```

You are capturing the pointer too soon, you must read the length first then do sectionName = m_pc.
Right now, the name string will include the leb128 encoded length

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 179 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1460#-KSEFBTXzF5Omt9sRowi:-KSEFBTYgiWFJfsEfQjc:b-62veos) ([raw file](https://github.com/microsoft/chakracore/blob/6d9b004f8e728e70a1432dce15e06a59dba65a79/lib/WasmReader/WasmBinaryReader.cpp#L179)):*
> ```cpp
>         sectionName = (char*)(m_pc); //including "name" section
>         nameLength = LEB128(len);
>         m_pc += nameLength; //skip section name for now
> ```

Should add a check here `CheckBytesLeft(nameLength);` before moving the pointer.

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 182 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1460#-KSEEQX2u3a8lNSWIx4d:-KSEEQX2u3a8lNSWIx4e:b-casx1n) ([raw file](https://github.com/microsoft/chakracore/blob/6d9b004f8e728e70a1432dce15e06a59dba65a79/lib/WasmReader/WasmBinaryReader.cpp#L182)):*
> ```cpp
>     }
> 
>     sectionSize = LEB128(len);
> ```

According to [Section Header](https://github.com/WebAssembly/design/blob/binary_0xc/BinaryEncoding.md#high-level-structure)
The sectionSize comes before the name and after the id.
Also, the end is calculated with `payload_len - sizeof(name) - sizeof(name_len)`

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 187 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1460#-KSEFQqheL7MgCcINXfB:-KSEFQqheL7MgCcINXfC:bcm6itf) ([raw file](https://github.com/microsoft/chakracore/blob/6d9b004f8e728e70a1432dce15e06a59dba65a79/lib/WasmReader/WasmBinaryReader.cpp#L187)):*
> ```cpp
> 
> #if ENABLE_DEBUG_CONFIG_OPTIONS
>     Assert(nameLength < 64);
> ```

While your at it, instead of an assert, could you make this a check and just print Unknown section name if the length is greater than 64.

---

*[test/rlexedirs.xml, line 376 at r2](https://reviewable.io:443/reviews/microsoft/chakracore/1460#-KSEFovvYzTdM4k5L4C2:-KSEFovvYzTdM4k5L4C3:bola50p) ([raw file](https://github.com/microsoft/chakracore/blob/f1e29a11282c65950b528582d2a0bd96a3305562/test/rlexedirs.xml#L376)):*
> ```XML
> <dir>
>   <default>
>     <files>WasmSpec</files>
> ```

This is going to conflict with George's change.
I suggest to wait until he merges it than resolve this file using his version since I prefer to comment out these tests for the time being rather than deleting it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1460)*

**Comment by Krovatkin** [comment]
@Cellule that's amazing how much I managed to mess up in like 10 lines of code :-)))))) Regarding removing the tests George and I weren't quite sure which change goes first. Apparently his does :-) I'll back out that part and fix the rest.

**Comment by Krovatkin** [comment]
@Cellule updated the PR.

**Comment by Cellule** [comment]
Reviewed 9 of 10 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 178 at r3](https://reviewable.io:443/reviews/microsoft/chakracore/1460#-KSgvaAE1xqkyQxL6Amu:-KSgvaAE1xqkyQxL6Amv:bdg1ecb) ([raw file](https://github.com/microsoft/chakracore/blob/913784825b9b509e99b878ad2f2aa105ff9bd882/lib/WasmReader/WasmBinaryReader.cpp#L178)):*
> ```cpp
>     }
>     else
>     {
> ```

What happens to `header.code` in this scenario ?
I assume it keeps it default value, should we have a value that means `user section` ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1460)*

**Comment by Krovatkin** [comment]
*[lib/WasmReader/WasmBinaryReader.cpp, line 178 at r3](https://reviewable.io:443/reviews/microsoft/chakracore/1460#-KSgvaAE1xqkyQxL6Amu:-KSln16DveOSKNMtSSPs:b34yqow) ([raw file](https://github.com/microsoft/chakracore/blob/913784825b9b509e99b878ad2f2aa105ff9bd882/lib/WasmReader/WasmBinaryReader.cpp#L178)):*
<details><summary><i>Previously, Cellule (Michael Ferris) wrote…</i></summary>
> What happens to `header.code` in this scenario ?
> I assume it keeps it default value, should we have a value that means `user section` ?
</details>
hhmmm, it is unlikely that  it gets a default value, since its parent structure starts out as an uninitialized struct :-( I think adding an explicit user section is a really great idea. I updated the PR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1460)*

**Comment by Cellule** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/1460#-:-KSm9SxCd6GST2ui8an_:bnfp4nl)*

**Comment by Krovatkin** [comment]
@arunetm Could you please merge this when you have a mnt?

**Comment by arunetm** [comment]
Merged to WebAssembly branch.

**Pull request merged**